### ISSUE TITLE
Bug in RunProperties.IsBold()

### ISSUE THREAD

**Issue by preciselytom** [baliance/gooxml#204]
This is the method code:
```
func (r RunProperties) IsBold() bool {
	return r.x.B != nil
}
```
It works in most cases, since a non-bold run's properties look like this:
```
            B: (*wml.CT_OnOff)(<nil>),
```
And a bold run's properties look like this:
```
                B: (*wml.CT_OnOff)(0xc42000e450)({
                 ValAttr: (*sharedTypes.ST_OnOff)(<nil>)
                }),
```
However, if the style uses bold by default, and a run turns it off, then B will look like this:
```
                B: (*wml.CT_OnOff)(0xc42000e3a0)({
                 ValAttr: (*sharedTypes.ST_OnOff)(0xc420011d80)(false)
                })
```
Therefore, `IsBold()` would falsely say that the run is bold.

This seems to map directly from XML where `<w:b/>` is short for `<w:b val="true"/>`, and turning bold off requires `<w:b val="false"/>`.

Actually I can't say what the proper behaviour should be, because a boolean is not enough to express the three possible states:
* No change to boldness
* Turn on boldness
* Turn off boldness

Perhaps two methods are needed: `IsBold()` and `BoldSet()`:

```
func (r RunProperties) IsBold() bool {
	if r.x.B != nil {
		if r.x.B.ValAttr != nil && r.x.B.ValAttr.Bool != nil && *r.x.B.ValAttr.Bool == false {
			return false
		} else {
			return true
		}
	} else {
		return false
	}
}
func (r RunProperties) BoldSet() bool {
	return r.x.B != nil
}
```

I also noticed an oddity in `IsItalic()`:
```
func (r RunProperties) IsItalic() bool {
	if r.x == nil {
		return false
	}
	return r.x.I != nil
}
```
All other methods assume that `r.x` is safely non-nil, so why this?

**Comment by tbaliance** [comment]
Thanks, I pushed up a change with a corrected IsBold method.  Check it out and let me know and I'll get it merged in and tagged soon.  

What would you use your IsBoldSet for? I added it at first, but couldn't come up with a good use case for the comment.

**Comment by preciselytom** [comment]
Looks good! But it still doesn't distinguish between the case where boldness is left as-is (on or off), and the case where boldness is turned off in case it was set by a style further down the hierarchy. My proposed method `IsBoldSet()` would tell whether the RunProperties modifies boldness.

AFAIK, in terms of XML, there are three ways that a `<w:rPr>` can deal with boldness:
1. `<w:b/>` turns on boldness
2. `<w:b val="false"/>` turns off boldness
3. Absence of `w:b` means that boldness is left as is

**Comment by tbaliance** [comment]
@preciselytom And you've got the same problem for Italic right?  I think it works the same way.

**Comment by preciselytom** [comment]
Yes, I think it applies to any boolean property.

Actually this issue might be on a higher level. Gooxml doesn't seem to include methods for calculating the effective run and paragraph properties at a given point in the document. I'm implementing such a feature right now, in the software where we use gooxml. It involves walking the document tree (using the wml types) recursively, applying properties on the way.

**Comment by tbaliance** [comment]
I came up with another way to just directly return the three values with BoldValue/ItalicValue methods rather than users needing to check two boolean, please have a look.

Yes, gooxml doesn't have any support for calculating computed styles.  Where we've used it, we've been mostly generating documents from scratch and haven't had the need. Are you generating, or editing existing documents?

**Comment by preciselytom** [comment]
Thanks, I like this implementation even better.

We are converting between Word files and our own (simpler) document format, which has only direct formatting, so we have to calculate the effective inherited formatting in Word, and turn it into direct formatting. I started using Document.Paragraphs() but switched to wml structs when I realized that the order of paragraphs and tables is not preserved with those methods.
